### PR TITLE
Add the Element Origin concept

### PR DESCRIPTION
**Pull request by jgraham**
This is really designed for BiDi, allowing actions to specify the
origin relative to an element in the form:

 {
  type: "element"
  sharedId: <id>
 }

rather than sending in a WebDriver-style element reference. For
simplicit this allows the new construct to work irrespective of
whether actions are being used in a HTTP session or a BiDi
session (but a BiDi session is not expected to support the HTTP-style
web element reference identifiers).

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/pull/1653.html" title="Last updated on Dec 22, 2022, 4:35 PM UTC (24e4604)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/1653/6c3b67a...24e4604.html" title="Last updated on Dec 22, 2022, 4:35 PM UTC (24e4604)">Diff</a>

### CODE REVIEW

File: index.html
Comment (whimboo) [inline index.html:7720]: In the BiDi spec we have `type=node` when serializing DOM nodes. So I assume that we should have that here as well?

Comment (jgraham) [reply index.html]: Yeah, although given this should really be an Element and not something else, I wonder if we should actually give Element specific serialization that differs from Node. I'll file an issue.

File: index.html
Comment (whimboo) [inline index.html:7766]: Nothing in the current WebDriver classic spec actually uses `sharedId`. So if only classic is used any retrieved and serialized node will not be added to the `list of known elements` containing a `sharedId` property.

I assume that this PR is blocked by some other PR. You might want to mention that in the initial comment and probably we should add a new label to indicate that? Right now dependencies are not visible.

File: index.html
Comment (whimboo) [inline index.html:7765]: I think we should update that part to make use of the window scope (depending on which will land earlier). See PR #1667 for the upcoming changes.

File: index.html
Comment (jgraham) [reply index.html:7766]: The intent is that by getting the `sharedId` property from the object you're effectively extracting the web element reference identifier. Other parts of the spec shouldn't care that it came from a BiDi style `{sharedId: <id>}` object vs a class-style webelement reference object.

File: index.html
Comment (jgraham) [reply index.html:7765]: This isn't perfectly sorted, but I don't think it should block this.

**Comment by jgraham** [comment]
@foolip @sadym-chromium could one of you review this please?

File: index.html
Comment (foolip) [inline index.html:7757]: I'm a bit unclear about what exactly this check is. Following the definition of represents a web element, would this match for the object `{ "element-6066-11e4-a52e-4f735466cecf": "whatever" }`?

It would be good to spell out what the element origin type test looks like, in particular if we tolerate additional keys in the map beyond "type" and "sharedId".

**Review by sadym-chromium — COMMENTED** [review]
Still some investigation for processing `sharedId` in Classic form our side is needed.

File: index.html
Comment (whimboo) [reply index.html:7767]: I think that this relies on https://github.com/w3c/webdriver-bidi/issues/236 where we want to add a `stale` like field to the serialized data.

Comment (sadym-chromium) [reply index.html:7767]: makes sense

**Comment by sadym-chromium** [comment]
So the suggested way is attractive, but I guess we should not extend the Classic protocol with BiDi format. To avoid that, we can pass a `deserialization algorithm`, or more hacky boolean `elementIdField`.

**Comment by foolip** [comment]
Passing in an algorithm to deserialize seems fine to me. @jgraham WDYT?

**Comment by jgraham** [comment]
Sure, if we don't want to just allow the new format in classic then passing in the deserialization steps is a pretty clean approach.

**Comment by sadym-chromium** [comment]
> Sure, if we don't want to just allow the new format in classic then passing in the deserialization steps is a pretty clean approach.

`is element origin` and `get element origin` solves it. LGTM

**Review by sadym-chromium — APPROVED** [review]

**Review by whimboo — CHANGES_REQUESTED** [review]
@jgraham the checks are still failing given some wrong references which would need an update:

https://github.com/w3c/webdriver/actions/runs/3732469018/jobs/6331952282#step:3:893

**Review by whimboo — APPROVED** [review]
This looks all fine to me as well. Please check the 3 comments before merging. Thanks!

**Comment by whimboo** [comment]
@jgraham mind updating this PR so that we can get it merged?

**Pull request merged**
